cli/cli_app.go

```diff
@@ -264,6 +266,57 @@ func shouldPrintTerraformHelp(terragruntOptions *options.TerragruntOptions) bool
 	return false
 }
 
+// loadEnvVars loads environment variable definitions from a shell command.
+// Currently the function expects KEY=value output to STDOUT
+// the method returns the old evironment variables map (dictionary) if the
+// changes should be transient.
+func loadEnvVars(terragruntOptions *options.TerragruntOptions, curHook config.Hook) (map[string]string, error) {
+	var oldEnvMap map[string]string
+
+	if curHook.LoadEnvVars != nil {
+		terragruntOptions.Logger.Printf("Loading environment for hook: %s", curHook.Name)
+
+		actionToExecute := curHook.LoadEnvVars.Execute[0]
+		actionParams := curHook.LoadEnvVars.Execute[1:]
+		out, err := shell.RunShellCommandWithOutput(terragruntOptions, "", actionToExecute, actionParams...)
+		if err != nil {
+			terragruntOptions.Logger.Printf("Error running hook %s while loading environment variables with message: %s", curHook.Name, err.Error())
+			return nil, err
+		}
+		// parse out from script to map
+		m := make(map[string]string)
+		scanner := bufio.NewScanner(strings.NewReader(out.Stdout))
+		for scanner.Scan() {
+			txt := scanner.Text()
+			idx := strings.Index(txt, "=")
+			if idx > 0 {
+				key := strings.TrimSpace(txt[0:idx])
+				value := txt[idx+1 : len(txt)]
+				m[key] = value
+			} else {
+				terragruntOptions.Logger.Print("Invalid key,value pair")
+			}
+		}
+		targetEnvMap := terragruntOptions.Env
+		if curHook.LoadEnvVars.Transient {
+			oldEnvMap = terragruntOptions.Env
+			targetEnvMap := make(map[string]string)
+			for key, value := range terragruntOptions.Env {
+				targetEnvMap[key] = value
+			}
+			terragruntOptions.Env = targetEnvMap
+		}
+		if curHook.LoadEnvVars.Overwrite {
+			targetEnvMap = make(map[string]string)
+			terragruntOptions.Env = targetEnvMap
```

Comment:
I'm not a fan of modifying `terragruntOptions` in place. It makes the code hard to follow, requires the user of this function to know to retain the old map of env vars and reset it later, and may cause issues in concurrent situations (e.g., if we have several goroutines using the same `terragruntOptions`). Instead, I'd recommend that this method returns a _new_ `terragruntOptions`, cloned from the old one, but with env vars overwritten. That way, there are no concurrency concerns, and the caller of the method can use the new `terragruntOptions` with no concern about having to reset anything in the old one.

Reply:
The argument about potential concurrency issues is a good one and convinced me. Although I'M not a big fan of bloating the memory with most of the useless copies. Let's hope the GC of GO is fast and clever enough :-D

Reply:
Perf is not going to be a concern at all in a tool like Terragrunt. Even if it was, we'd still optimize for correctness/maintenance/readability/ first, then find the bottlenecks, and optimize solely the code relevant to those.